### PR TITLE
feat(frontend): UX change to support downloading directory artifacts. Fixes #3667

### LINKED ISSUE #3667: Unable to download directory artifact from the UX link
### What steps did you take:
I am using the off-the-shelf component to download a directory from GCS. It seems to run to completion, but when I attempt to download the artifact from minio, it fails.

here is the pipeline:

```shell
import_op = components.load_component_from_url(
    'https://raw.githubusercontent.com/kubeflow/pipelines/2dac60c/components/google-cloud/storage/download_dir/component.yaml'

def trainer(train_path: 'URI' = 'gs://my-fake-bucket/tf_records/*'):
  import_train_data_task = import_op(train_path)
```

### What happened:
Error when trying to download

### What did you expect to happen:
Successfully able to download files

### Environment:
Hosted KFP

How did you deploy Kubeflow Pipelines (KFP)?

KFP version: 

KFP SDK version: 

### Anything else you would like to add:
[Miscellaneous information that will assist in solving the issue.]

/kind bug

### PR DESCRIPTION
**Pull request by Bobgy**
**Description of your changes:**
* Added a new UI server endpoint: `/artifacts/:source/:bucket/:key`, e.g. `/artifacts/minio/mlpipeline/hello.tgz`. This endpoint downloads artifacts without trying to extract from `*.tar.gz`, so it can download tar gzed directory artifacts.
* Changed artifact url to point to new added download url.
* When artifact preview cannot show all content, show a `View All` button below with original full text view url.

~~**DEMO: https://youtu.be/IDcxRLr-kcw**~~
**Updated DEMO: https://youtu.be/iPu6v5n5onI**

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by Ark-kun** [comment]
Some thoughts:
* Gzipping is an implementation detail of Argo. We had a PR to disable this https://github.com/kubeflow/pipelines/pull/1640 The only reason we did not disable the compression yet is because there were concerns about the UX breaking. If you're fixing the UX, maybe we can make sure the uncompressed artifacts work in the UX and turn odd the compression.
* The only case where the user might want to download a .tar.gz archive is when the artifact is a directory. This represents a minority of the artifact use-cases. Let's try to ensure the majority scenario does not degrade.

**Comment by Ark-kun** [comment]
Currently the "Show All" link is in the bottom. Can you show it at the top alongside the archive link? I think we do not need to show the actual archive URI.

```
some_output: [View] [Download] [Visualize] 37KiB
[preview]
```

**Comment by Bobgy** [comment]
> Currently the "Show All" link is in the bottom. Can you show it at the top alongside the archive link? I think we do not need to show the actual archive URI.
> 
> ```
> some_output: [View] [Download] [Visualize] 37KiB
> [preview]
> ```

I think the actual archive URI is still useful for cases like using GCS as the storage.

**Comment by Bobgy** [comment]
Hi @eterna2, do you want to review this UX change?
Feedback welcomed on the DEMO, you can decide to review code or not, there's a fair amount of them.

**Comment by eterna2** [comment]
> Hi @eterna2, do you want to review this UX change?
> Feedback welcomed on the DEMO, you can decide to review code or not, there's a fair amount of them.

Sure. I will have a look at it tomorrow.

**Comment by Ark-kun** [comment]
/approve cancel
(to prevent auto-submission)

**Comment by Ark-kun** [comment]
/lgtm

There seems to be a recent change where PRs started being auto-approved.
Please unhold when you are ready.
/hold

**Comment by Bobgy** [comment]
Thanks @Ark-kun! I'll wait for @eterna2's opinion before drawing a final conclusion.

File: frontend/server/minio-helper.ts [hunk 1]
```diff
@@ -126,13 +128,18 @@ function extractFirstTarRecordAsStream() {
...
+ * @param param.tryExtract Whether we try to extract *.tar.gz, default to true.
  *
  */
 export async function getObjectStream({
   bucket,
   key,
   client,
-}: MinioRequestConfig): Promise<Transform> {
+  tryExtract,
+}: MinioRequestConfig & { extract?: boolean }): Promise<Transform> {
```
Comment (eterna2) [inline frontend/server/minio-helper.ts:139]: `{ extract?: boolean }` what is this line suppose to do? I don't see `extract` used anywhere? i.e. u already modified `MinioRequestConfig` with `tryExtract`

File: frontend/server/minio-helper.ts [hunk 2]
```diff
@@ -126,13 +128,18 @@ function extractFirstTarRecordAsStream() {
...
   const stream = await client.getObject(bucket, key);
-  return stream.pipe(gunzip()).pipe(maybeTarball());
+  if (tryExtract == null) {
```
Comment (eterna2) [inline frontend/server/minio-helper.ts:141]: u can try doing this to set the default.
```ts
export async function getObjectStream({
  bucket,
  key,
  client,
  tryExtract=true,
}: MinioRequestConfig): Promise<Transform> {
  ...
}
```

**Comment by Bobgy** [comment]
Thank you for catching that @eterna2 ! Fixed the two problems

**Pull request merged**